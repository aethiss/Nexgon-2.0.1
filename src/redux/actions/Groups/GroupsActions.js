import uid from 'uid-safe'
import { _setStorageItem } from '../../../libs/Storage/AsyncStorage'
import { createNewGroupMessages } from './MessagesActions'

function deleteGroup(group) {
  return {
    type: 'REMOVE_GROUP',
    group,
  }
}

export function addGroup(group) {
  return {
    type: 'ADD_GROUP',
    group,
  }
}


// ====== HANDLERS ======= //

export function createNewGroup(data) {
  return (dispatch, getState) => {
    let error = false

    // validate group object
    // eslint-disable-next-line array-callback-return
    Object.keys(data).map((key) => {
      if (data[key] === undefined || data[key] === '') {
        // TODO: Dispatch global errors
        error = { error: `Group Not Valid : ${key}` }
      }
    })

    if (!error) {
      const newGroup = {
        notification: 0,
        groupId: uid.sync(18),
        created_at: Date.now(),
        ...data,
      }
      dispatch(addGroup(newGroup))
      _setStorageItem('groups', getState().groups)

      // Creiamo una chat vuota e salviamo nel local store (TODO: da sistemare)
      dispatch(createNewGroupMessages(newGroup.groupId))
      _setStorageItem('messages', getState().messages)

      return { result: 'group created' }
    }
    return error
  }
}

export function removeGroup(group) {
  return (dispatch, getState) => {
    dispatch(deleteGroup(group))
    _setStorageItem('groups', getState().groups)
    // TODO: remove/update messages group
  }
}
