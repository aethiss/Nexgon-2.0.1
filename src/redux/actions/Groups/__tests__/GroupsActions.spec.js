/* eslint-disable no-undef */
import configureStore from '../../../lib/ConfigureStore'
import * as actions from '../GroupsActions'

const snapshot = action => expect(action).toMatchSnapshot()

describe('[Actions Groups]', () => {
  it('Delete Group)', () => {
    const store = configureStore()
    const group1 = {
      groupId: '1234567890ABCDEFGH',
      name: 'ZeroOne',
      visible: true,
      inviteOnly: false,
      notification: 2,
      image: 'https://unsplash.it/150/?random',
      create_at: Date.now(),
    }
    const group2 = {
      groupId: '0987654321GFDSA',
      name: 'Amici',
      visible: false,
      inviteOnly: true,
      notification: 5,
      image: 'https://unsplash.it/150/?random',
      create_at: Date.now(),
    }
    const group = {
      groupId: '0987654321GFDSA',
      name: 'Amici',
      visible: false,
      inviteOnly: true,
      notification: 5,
      image: 'https://unsplash.it/150/?random',
      create_at: Date.now(),
    }
    snapshot(actions.removeGroup(group))
    store.dispatch(actions.createNewGroup(group1))
    store.dispatch(actions.createNewGroup(group2))
    store.dispatch(actions.removeGroup(group))
    expect(store.getState().groups.length).toBe(1)
  })

  it('Create New Group', () => {
    const store = configureStore()
    const newgroup = {
      name: 'Amici',
      visible: false,
      inviteOnly: true,
      image: 'https://unsplash.it/150/?random',
    }
    snapshot(actions.createNewGroup(newgroup))
    store.dispatch(actions.createNewGroup(newgroup))
    expect(store.getState().groups.length).toBe(1)
    expect(store.getState().groups[0].notification).toBe(0)
  })
  it('Error on group creation', () => {
    const store = configureStore()
    const newgroup = {
      name: '',
      visible: false,
      inviteOnly: true,
      image: 'https://unsplash.it/150/?random',
    }
    snapshot(actions.createNewGroup(newgroup))
    store.dispatch(actions.createNewGroup(newgroup))
    expect(store.getState().groups.length).toBe(0)
  })
})
