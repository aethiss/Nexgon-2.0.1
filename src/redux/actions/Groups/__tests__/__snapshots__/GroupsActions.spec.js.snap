// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Actions Groups] Create New Group 1`] = `[Function]`;

exports[`[Actions Groups] Delete Group) 1`] = `[Function]`;

exports[`[Actions Groups] Error on group creation 1`] = `[Function]`;
